src/utils/sized_buffer.rs

```diff
@@ -0,0 +1,72 @@
+use std::io::{Error as IoError, ErrorKind, Write};
+
+pub(crate) struct SizedBuffer {
+    inner: Vec<u8>,
+    limit: usize,
+}
+
+impl SizedBuffer {
+    pub(crate) fn new(limit: usize) -> Self {
+        SizedBuffer {
+            inner: Vec::new(),
+            limit,
+        }
+    }
+
+    pub(crate) fn reserve(&mut self, amount: usize) {
+        if self.inner.len() + amount > self.limit {
+            self.inner.reserve_exact(self.limit - self.inner.len());
+        } else {
+            self.inner.reserve_exact(amount);
+        }
+    }
```

Comment:
Maybe this should return the amount that it ended up reserving so that we can pivot off of that as needed, allowing us to maybe less-fatally operate by retrying writes after the buffer's flushed

Reply:
Do you see this being needed anywhere in the current PR? If future code needs it, the return value can be added in a future PR.